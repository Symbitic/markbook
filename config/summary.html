<!DOCTYPE HTML>
<html>
  <head>
    <!-- Generated using Markbook -->
    <meta charset="UTF-8">
    <title>Markbook - SUMMARY.md</title>
    <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#ffffff" />
    <base href="..">
    <link rel="shortcut icon" href="favicon.png">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/bulma/0.7.4/css/bulma.min.css" integrity="sha256-8B1OaG0zT7uYA572S2xOxWACq9NXYPQ+U5kHPV1bJN4=" crossorigin="anonymous">
    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.2/css/all.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
    <link rel="stylesheet" href="markbook.css">
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/3.4.4/fuse.min.js" integrity="sha256-9tlwNm1VkLGYfwoO2fFKKVK/mGmaARt76rM7hLUkaeY=" crossorigin="anonymous"></script>
  </head>
  <body>
    <div class="columns is-mobile is-vcentered">
      <!-- Show/Hide Searchbar -->
      <div class="column is-narrow">
        <span aria-controls="searchbar" aria-expanded="false" aria-label="Toggle Searchbar" class="icon is-medium">
          <span aria-hidden="true" class="fas fa-search"></span>
        </span>
      </div>
      
      <!-- Chapter Title -->
      <div class="column is-wrapped has-text-centered">
        <p class="title has-wrapped-overflow has-text-grey-lighter">
          Markbook
        </p>
      </div>

      <!-- Show/Hide Table of Contents -->
      <div class="column has-text-left is-narrow">
        <span aria-controls="sidebar" aria-expanded="true" aria-label="Toggle Table of Contents" class="icon is-medium">
          <span aria-hidden="true" class="fas fa-bars"></span>
        </span>
      </div>
    </div>
    
    <section class="section">
      <div class="columns">
        <div id="sidebar" class="column is-2 is-paddingless is-info">
          <aside class="is-medium menu" aria-label="Table of Contents">
            <ul class="menu-list">
              <li><a href="index.html">Markbook</a></li>
              <li><a href="cli/index.html">
                <strong aria-hidden="true">1.</strong>
                Command Line Tool
              </a></li>
              <li><a href="cli/build.html">
                <strong aria-hidden="true">1.1</strong>
                build
              </a></li>
              <li><a href="config/index.html">
                <strong aria-hidden="true">2.</strong>
                Configuration
              </a></li>
              <li><a href="config/summary.html">
                <strong aria-hidden="true">2.1</strong>
                SUMMARY.md
              </a></li>
              <li><a href="writing/index.html">
                <strong aria-hidden="true">3.</strong>
                Writing
              </a></li>
              <li><a href="writing/commonmark.html">
                <strong aria-hidden="true">3.1</strong>
                CommonMark
              </a></li>
              <li><a href="writing/extensions.html">
                <strong aria-hidden="true">3.2</strong>
                Markbook-Specific Extensions
              </a></li>
              <li><a href="contributors.html">Contributors</a></li>

            </ul>
          </aside>
        </div>
        <div id="content" class="column is-10">
          <div class="content is-medium">
            <div class="container">
              <input id="searchbar" class="input is-rounded is-small hidden" type="text" placeholder="Search...">
              <div id="search-results"></div>
              <h1>SUMMARY.md</h1>
            <p>The summary file is used by markbook to know what chapters to include, in what order they should appear, what their hierarchy is and where the source files are. Without this file, there is no book.</p>
            <p>Even though <code>SUMMARY.md</code> is a markdown file, the formatting is very strict to allow for easy parsing. Let's see how you should format your <code>SUMMARY.md</code> file.</p>
            <h2>Format</h2>
            <ol>
            <li>
            <p><strong><em>Title</em></strong> It's common practice to begin with a title, generally # Summary. But it is not mandatory, the
            parser just ignores it. So you can too if you feel like it.</p>
            </li>
            <li>
            <p><strong><em>Prefix Chapter</em></strong> Before the main numbered chapters you can add a couple
            of elements that will not be numbered. This is useful for forewords,
            introductions, etc. There are however some constraints. You can not nest
            prefix chapters, they should all be on the root level. And you can not add
            prefix chapters once you have added numbered chapters.</p>
            <pre><code class="language-markdown">[Title of prefix element](relative/path/to/markdown.md)
            </code></pre>
            </li>
            <li>
            <p><strong><em>Numbered Chapter</em></strong> Numbered chapters are the main content of the book,
            they will be numbered and can be nested, resulting in a nice hierarchy
            (chapters, sub-chapters, etc.)</p>
            <pre><code class="language-markdown">-   [Title of the Chapter](relative/path/to/markdown.md)
            </code></pre>
            <p> You can either use <code>-</code> or <code>*</code> to indicate a numbered chapter.</p>
            </li>
            <li>
            <p><strong><em>Suffix Chapter</em></strong> After the numbered chapters you can add a couple of
            non-numbered chapters. They are the same as prefix chapters but come after
            the numbered chapters instead of before.</p>
            </li>
            </ol>
            <p>All other elements are unsupported and will be ignored at best or result in an error.</p>
            </div>
          </div>
        </div>
      </div>
    </section>
    
    <script src="search.js"></script>
    <script src="markbook.js"></script>
  </body>
</html>
