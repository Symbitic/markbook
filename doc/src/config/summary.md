# SUMMARY.md

The summary file is used by markbook to know what chapters to include, in what order they should appear, what their hierarchy is and where the source files are. Without this file, there is no book.

Even though `SUMMARY.md` is a markdown file, the formatting is very strict to allow for easy parsing. Let's see how you should format your `SUMMARY.md` file.

## Format

1. **_Title_** It's common practice to begin with a title, generally `# Summary`. It is not mandatory; the parser just ignores it.

2. **_Prefix Chapter_** Before the main numbered chapters you can add a couple of elements that will not be numbered. This is useful for forewords, introductions, etc. There are however some constraints. You can not nest prefix chapters, they should all be on the root level. And you can not add prefix chapters once you have added numbered chapters.

    ```markdown
    [Title of prefix element](relative/path/to/markdown.md)
    ```

3. **_Numbered Chapter_** Numbered chapters are the main content of the book, they will be numbered and can be nested, resulting in a nice hierarchy (chapters, sub-chapters, etc.)

    ```markdown
    * [Title of the Chapter](relative/path/to/markdown.md)
    ```

    You can either use `-` or `*` to indicate a numbered chapter.

4. **_Suffix Chapter_** After the numbered chapters you can add a couple of non-numbered chapters. They are the same as prefix chapters but come after the numbered chapters instead of before.

All other elements are unsupported and will be ignored (at best) or result in an error.

## Example

Here is an example of the SUMMARY.md used for Markbook:

```markdown
[Markbook](README.md)

* [Command-Line Interface](cli/README.md)
    + [build](cli/build.md)
```
