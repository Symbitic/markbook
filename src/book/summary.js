/**
 * Yes. This is needlessly big and complex. And it isn't particularly
 * functional-style.
 *
 * Who's bright idea was it to use a markdown file to specify the
 * table-of-contents in a markdown file?
 *
 * @todo Might need to add assert that level is never bigger than 2.
 */

import { reject } from 'common/errors.js'
import markdown from 'remark-parse'
import path from 'path'
import unified from 'unified'
import vfile from 'to-vfile'

/**
 * Summary item.
 * @typedef {Object} SummaryItem
 * @property {string} name - Chapter name.
 * @property {string} filename - Path to the chapter's source file.
 */

/**
 * Summary object
 * @typedef {Object} Summary
 * @property {SummaryItem[]} prefix - Chapters before the main text (e.g. an introduction).
 * @property {SummaryItem[]} chapters - The main body of the book.
 * @property {SummaryItem[]} suffix - Chapters after the main text (e.g. a conclusion).
 */

/**
 * Split remark tree into headers, prefixes, lists, and suffixes
 * @todo Really need to make this more FP-style. Recursion, Object.assign, etc.
 * @todo Maybe delete all root non-heading/paragraph/list nodes.
 */
const split = tree => {
  let isDone = false
  let index = 0

  const headers = tree.children.filter((node, i) => {
    if (isDone) {
      return false
    } else if (node.type === 'heading') {
      return true
    } else {
      index = i
      isDone = true

      return false
    }
  })

  isDone = false

  const prefixes = tree.children.slice(index).filter((node, i) => {
    if (isDone) {
      return false
    } else if (node.type === 'paragraph') {
      return true
    } else {
      isDone = true
      index = index + i
      return false
    }
  })

  isDone = false

  const lists = tree.children.slice(index).filter((node, i) => {
    if (isDone) {
      return false
    } else if (node.type === 'list') {
      return true
    } else {
      isDone = true
      index = index + i
      return false
    }
  })

  isDone = false

  const suffixes = tree.children.slice(index).filter((node, i) => {
    if (isDone) {
      return false
    } else if (node.type === 'paragraph') {
      return true
    } else {
      isDone = true
      index = index + i
      return false
    }
  })

  return {
    headers,
    prefixes,
    lists,
    suffixes
  }
}

/**
 * ES6 compose function.
 */
const compose = (...fns) =>
  fns.reduceRight((f, g) => (...args) => f(g(...args)))

/**
 * Method for flattening nested arrays. Intended for .reduce()
 */
const flattener = (acc, val) =>
  Array.isArray(val) ? acc.concat(val.reduce(flattener, [])) : acc.concat(val)

/**
 * Remove unneeded properties from links arrays.
 */
const removeExtra = items =>
  items.map(item => {
    const { type, children, url, value } = item
    return JSON.parse(
      JSON.stringify({
        type,
        url,
        value,
        children: Array.isArray(children) ? removeExtra(children) : children
      })
    )
  })

/**
 * Extract the header.
 */
const extractHeader = headers =>
  headers
    .map(node => node.children)
    .reduce(flattener, [])
    .map(node => (node.value ? node.value : ''))[0] // eslint-disable-line no-unexpected-multiline

/**
 * Extract the prefix.
 */
const extractPrefix = items =>
  items
    .map(node => node.children.filter(node => node.type === 'link'))
    .reduce(flattener, [])
    .map(node => ({
      title: node.children.reduce(
        (acc, val) => (val.type === 'text' ? acc.concat(val.value) : acc),
        ''
      ),
      url: node.url
    }))

/**
 * Extract the suffix.
 */
const extractSuffix = items =>
  items
    .map(node => node.children.filter(node => node.type === 'link'))
    .reduce(flattener, [])
    .map(node => ({
      title: node.children.reduce(
        (acc, val) => (val.type === 'text' ? acc.concat(val.value) : acc),
        ''
      ),
      url: node.url
    }))

/**
 * Find paragraph nodes and add them (including a nested level indicator).
 */
const findParagraphs = (items, i = 0) =>
  items.reduce((acc, val) => {
    if (val.type === 'paragraph') {
      return acc.concat({
        ...val,
        level: i / 2
      })
    } else {
      return acc.concat(findParagraphs(val.children, i + 1))
    }
  }, [])

/**
 * Flatten the links array to just what is needed.
 */
const flattenLinks = items =>
  items.reduce((acc, val) => {
    if (
      val.children &&
      val.children.length === 1 &&
      val.children[0].type === 'link'
    ) {
      return acc.concat({
        title: val.children[0].children[0].value,
        url: val.children[0].url,
        level: val.level
      })
    } else if (Array.isArray(val.children)) {
      return acc.concat(flattenLinks(val.children))
    } else {
      return acc.concat(val)
    }
  }, [])

/**
 * Remove anything that isn't a link item from this array.
 */
const linksOnly = items =>
  items.filter(
    item => item.title && item.title.length && item.url && item.url.length
  )

/**
 * Extract the different chapters.
 */
const extractChapters = compose(
  removeExtra,
  findParagraphs,
  flattenLinks,
  linksOnly
)

/**
 * Read a file (wrapped for better reject messages).
 */
const readFile = filename =>
  vfile.read(filename).catch(() => reject('SUMMARY.md not found', __filename))

/**
 * Parse SUMMARY.md (wrapped for better reject messages).
 */
const parse = file =>
  Promise.resolve(file)
    .then(file => {
      const tree = unified()
        .use(markdown)
        .parse(file)

      const { headers, prefixes, lists, suffixes } = split(tree)

      const header = extractHeader(headers)
      const prefix = extractPrefix(prefixes)
      const chapters = extractChapters(lists)
      const suffix = extractSuffix(suffixes)

      return {
        header,
        prefix,
        chapters,
        suffix
      }
    })
    .catch(() => reject('Error parsing SUMMARY.md'))

/**
 * Load and parse a SUMMARY.md file.
 * @param {!string} filename - Path to `SUMMARY.md`.
 * @returns {Summary} An object representing the books contents.
 */
export default function(config) {
  const summary = path.join(config.source, 'SUMMARY.md')

  return readFile(summary)
    .then(parse)
    .then(summary => ({
      ...config,
      summary
    }))
}
