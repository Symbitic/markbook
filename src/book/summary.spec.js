/* eslint-env jest */
import path from 'path'
import summary from './summary'

const case1 = ['parse a basic summary', path.join('tests', 'basic', 'src')]

const case2 = [
  'parse even complex summaries',
  path.join('tests', 'summary', 'complex')
]

const tests = [case1, case2]

describe.each(tests)('book/summary', (name, source) => {
  it(`should ${name}`, () => {
    return expect(summary({ source })).resolves.toMatchSnapshot()
  })
})
