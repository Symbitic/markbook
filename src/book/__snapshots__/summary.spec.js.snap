// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`book/summary should parse a basic summary 1`] = `
Object {
  "source": "tests/basic/src",
  "summary": Object {
    "chapters": Array [
      Object {
        "level": 1,
        "title": "Chapter 1",
        "url": "chapter-1.md",
      },
      Object {
        "level": 1,
        "title": "Chapter 2",
        "url": "chapter-2.md",
      },
    ],
    "header": "Summary",
    "prefix": Array [
      Object {
        "title": "Basic",
        "url": "README.md",
      },
    ],
    "suffix": Array [],
  },
}
`;

exports[`book/summary should parse even complex summaries 1`] = `
Object {
  "source": "tests/summary/complex",
  "summary": Object {
    "chapters": Array [
      Object {
        "level": 1,
        "title": "Documentation",
        "url": "README.md",
      },
      Object {
        "level": 1,
        "title": "Chapter 2",
        "url": "chapter-1.md",
      },
      Object {
        "level": 1,
        "title": "Topic",
        "url": "topic.md",
      },
      Object {
        "level": 2,
        "title": "Sub Topic",
        "url": "sub-topic.md",
      },
      Object {
        "level": 2,
        "title": "Sub Topic 2",
        "url": "sub-topic.md",
      },
      Object {
        "level": 3,
        "title": "Micro topic",
        "url": "micro-topic.md",
      },
      Object {
        "level": 2,
        "title": "Sub Topic 3",
        "url": "sub-topic.md",
      },
    ],
    "header": "Summary",
    "prefix": Array [
      Object {
        "title": "Copyright",
        "url": "COPYRIGHT.md",
      },
      Object {
        "title": "Prologue",
        "url": "PROLOGUE.md",
      },
      Object {
        "title": "Authors",
        "url": "AUTHORS.md",
      },
    ],
    "suffix": Array [
      Object {
        "title": "Bibliography",
        "url": "bibliography.md",
      },
    ],
  },
}
`;
